### ISSUE TITLE
Overlaps() commutativity and atomic intervals

### ISSUE THREAD

**Issue by msarrel** [AlexandreDecan/python-intervals#13]
The overlaps() function seems to have some problem with commutativity and atomic intervals.

This example has a problem:

`I.closed(1,1).overlaps(I.openclosed(1,2))`
`False` (correct)
`I.openclosed(1,2).overlaps(I.closed(1,1))`
`True` (should be False)

But, this example succeeds:

`I.closed(0,1).overlaps(I.openclosed(1,2))`
`False` (correct)
`I.openclosed(1,2).overlaps(I.closed(0,1))`
`False` (correct)

There is a similar problem with union.

This example has a problem:

`I.closed(1,1) | I.openclosed(1,2)`
`(1,2]` (should be [1,2])
`I.openclosed(1,2) | I.closed(1,1)`
`(1,2]` (should be [1,2])

But, this example succeeds:

`I.closed(0,1) | I.openclosed(1,2)`
`[0,2]` (correct)
`I.openclosed(1,2) | I.closed(0,1)`
`[0,2]` (correct)

**Comment by AlexandreDecan** [comment]
Thanks for reporting those bugs! I'll have a look in a few minutes. The second bug (about unions) seems similar to the one I went into yesterday (see #12).

**Comment by AlexandreDecan** [comment]
Release 1.9.0 is now available on PyPI.

**Comment by msarrel** [comment]
Thank you so much.  The `IntervalDict` looks very useful.  It will replace some _ad hoc_ code I had previously written.  One feature that would be useful for me is to define addition, subtraction, etc. for `IntervalDict`.  That is if we have two `IntervalDict`s:

[1, 3] | [5, 7] = 1
and
[2, 4] | [6, 8] = 2

Their sum would be a single `Interval Dict` with three entries:

[1, 2} | [5, 7) = 1
[2, 3) | [6, 7) = 3
[3, 4] | [7, 8] = 2

Regards,

Marc

**Comment by AlexandreDecan** [comment]
Thanks for your feedback! I'll think about this. I cannot add support for "+", "-", "*", etc. because it is tightly coupled with numbers, and this library does not make any assumption on what are used as bounds in intervals. However, I can imagine a rather generic `d.combine(other_d, func=lambda x,y: x + y)` that will do the trick.

**Comment by msarrel** [comment]
The `d.combine()` approach would be sufficient for me.  Thanks.  Just to be clear, though, the `func` supplied in `d.combine()` would only apply to the values in the `IntervalDict` (the range), not the keys (the domain)?

**Comment by AlexandreDecan** [comment]
Yes, on the values, not on the keys. I'll open a new issue for this feature, so I won't forget about it :-D (I don't think I'll have time to work on this in the next few days)

**Comment by AlexandreDecan** [comment]
Just to be sure, when you wrote "apply to the values (the range) not the keys (the domain)", you mean the converse? (keys are intervals/ranges, values are what are stored along with intervals), right?

**Comment by AlexandreDecan** [comment]
FYI, `IntervalDict.combine` is now available from the master branch. It will be part of the next release (probably 1.10.0). I don't know when this release will be available, it will depends on the number of other changes I would like to put in.

**Comment by AlexandreDecan** [comment]
Release 1.10.0 is now available on PyPI.

**Comment by msarrel** [comment]
Thank you!  I’ll take a look at it as soon as I have a little time…

Marc
-=-
393A -- Ground System Architecture & Systems Engineering
+1 818 393-7786

« Il semble que la perfection soit atteinte non quand il n'y a plus rien à ajouter, mais quand il n'y a plus rien à retrancher. »
"It seems that perfection is attained not when there is nothing more to add, but when there is nothing more to remove."
Antoine de Saint Exupéry, 1939

On 2019-09-26, at 00:43, Alexandre Decan <notifications@github.com<mailto:notifications@github.com>> wrote:


Release 1.10.0 is now available on PyPI.
